include/onnxruntime/core/framework/op_kernel.h

```diff
@@ -94,80 +114,55 @@ class OpKernelContext {
     return p_ml_value ? p_ml_value->GetMutable<T>() : nullptr;
   }
 
+  virtual MLDataType InputType(int index) const = 0;
+  virtual MLDataType OutputType(int index) const = 0;
+
   // In the case that memory allocation has not been done for an output tensor,
   // The memory allocation will be done on-the-fly with given tensor shape.
   // Return nullptr if the output is an unused optional output.
-  Tensor* Output(int index, const TensorShape& shape);
-
-  const logging::Logger& Logger() const {
-    return *logger_;
-  }
-
-  int InputCount() const {
-    return static_cast<int>(kernel_->Node().InputDefs().size());
-  }
-
-  int ImplicitInputCount() const {
-    return static_cast<int>(kernel_->Node().ImplicitInputDefs().size());
-  }
-
-  int OutputCount() const {
-    return static_cast<int>(kernel_->Node().OutputDefs().size());
-  }
+  virtual Tensor* Output(int index, const TensorShape& shape) = 0;
 
   /**
    * return an allocator on device 0, with memtype of OrtMemTypeDefault
    *
    */
-  Status GetTempSpaceAllocator(AllocatorPtr* output) const;
+  virtual Status GetTempSpaceAllocator(AllocatorPtr* output) const = 0;
 
   /**
   Return the fence of current node's input.
   @param index The index of the input.
   @returns Point to the Fence of the input MLValue.
   It is null if the input MLValue doesn't have fence or the input is optional.
   */
-  Fence_t InputFence(int index) const;
+  virtual Fence_t InputFence(int index) const = 0;
 
   /**
   Return the fence of current node's implicit input.
   @param index The index of the implicit input.
   @returns Point to the Fence of the implicit input MLValue.
   It is null if the input MLValue doesn't have fence or the input is optional.
   */
-  Fence_t ImplicitInputFence(int index) const;
+  virtual Fence_t ImplicitInputFence(int index) const = 0;
 
   /**
   Return the fence of current node's output identifed by index.
   @param index The index of the output.
   @returns Point to the Fence of the output MLValue.
   It is null if the output MLValue doesn't have fence or the output is optional.
   */
-  Fence_t OutputFence(int index) const;
```

Comment:
I think the change of this file should be changing some APIs to allow overwriting, right? if yes, you may keep the original functions' order so that reviewing will be easier :)

Reply:
or @weixingzhang you may want to create a pure interface and have this OpKernelContext not changed but one implementation of your interface.